parent/pom.xml

```diff
@@ -97,6 +97,8 @@
     </mailingLists>
 
     <properties>
+        <maven.compiler.source>1.8</maven.compiler.source>
+        <maven.compiler.target>1.8</maven.compiler.target>
```

Comment:
wouldn't setting up release flag be better option? How is this going to guarantee that no SE 9+ APIs are used in the code to compiled?

Reply:
> wouldn't setting up release flag be better option? 

I think it wouldn't for following reasons combined:
- if release property is set, compiler-plugin uses it to call compiler regardless of actual JDK used, hence for JDK8 this happens:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project org.eclipse.persistence.asm: Fatal error compiling: invalid flag: --release
```
- not all children migrated to/enforce JDK > 1.8 yet.

Reply:
> * not all children migrated to/enforce JDK > 1.8 yet.

how is this useful for those who did then? What are those having default compile with `release 11` and additional compile step with `release 8` supposed to do to adopt this change?

Reply:
good point @lukasj . The release flag seems to be the better option if all children are building with jdk11. But I wonder if they do and/or are going to migrate?

Reply:
Well I think it's better to _'force'_ those relying on JDK 8 to move forward than those on JDK 11 already to move backward

Reply:
I have re-read the "Jakarta EE 9 Release Plan". To me it is not clear if the sources have to be built with jdk8 or with jdk11. My interpretation is it can be built with any jdk version>=8 .

Reply:
> how is this useful for those who did then?

There is no impact on them in such case. But using `m*.c*.release` would impact those that are not build-with-JDK9+ ready.

> What are those having default compile with `release 11` and additional compile step with `release 8` supposed to do to adopt this change?

Nothing, as they already configured compiler plugin in its configuration section or with properties (which override the ones from parent should some be defined).

Adding `m*.c*.source/target` will mainly serve informational role as I see it. Because - I guess, children already configure compiler plugin which by default uses source/target=1.6 if not commanded otherwise.

Using `--release` is not enough _to guarantee that no SE 9+ APIs are used in the code_, because [Cross-Compilation Options for javac](https://docs.oracle.com/en/java/javase/11/tools/javac.html#GUID-AEEC9F07-CB49-4E96-8BC7-BCC2C7F725C9) says so, and some other steps have to be done, beside setting that property.

So, having ```<maven.compiler.target>1.8</maven.compiler.target>```
could be JakartaEE declaration of class level, but I do not like idea of ```<maven.compiler.release>1.8</maven.compiler.release>``` now.

Reply:
> Using `--release` is not enough _to guarantee that no SE 9+ APIs are used in the code_, because [Cross-Compilation Options for javac](https://docs.oracle.com/en/java/javase/11/tools/javac.html#GUID-AEEC9F07-CB49-4E96-8BC7-BCC2C7F725C9__CROSS-COMPILATIONOPTIONSFORJAVAC-7D3D83C3) says so, and some other steps have to be done, beside setting that property.

This is false. On the same page, the description of `--release` reads:
> Compiles against the public, supported and documented API for a specific VM version.

There are 2 possibilities:
* the legacy way: `-source`+`-target`+`-bootclasspath`+`-extdirs`
* the modern way: `--release`

As @lukasj already pointed out, just setting `-source` and `-target` is not sufficient to prevent usage of Java SE 9+ APIs.

Reply:
> There are 2 possibilities:
>     * the legacy way: `-source`+`-target`+`-bootclasspath`+`-extdirs`
>     * the modern way: `--release`

Ouch, now I see. I was wrong. Thanks.
I agree - `release` is better :smile:  then.

Reply:
https://openjdk.java.net/jeps/247
I think we have the following options for the parent pom.xml:

1. Do nothing and abandon this PR. Handle everything in the child projects.
2. Add maven.compiler.release=8 to the parent pom and enforce building with jdk9+ in all child projects.
3. Add and activate two maven profiles for child projects building with either jdk8 or jdk9+.
For example:
3.1 profile for jdk9+ includes property release=8.
3.2 profile for jdk8 includes source=8, target=8.

I don't understand why -bootclasspath and -extdirs is needed in case 3.2. There is no cross compilation happening. (I am curious btw if option 3 was envisioned like this for the ecosystem.)

Reply:
> I don't understand why -bootclasspath and -extdirs is needed in case 3.2. There is no cross compilation happening.
because if you don't set these options, compiler may introduce a reference to a class that does not exist in the target JDK: this is one key thing that `--release` does = enforce classes from target JDK

one option often used to avoid setting -bootclasspath and -extdirs, but detect if something wrong happens is to use AnimalSniffer: https://www.mojohaus.org/animal-sniffer/